### PR TITLE
Make resizing raster-based again and offer "scale_method=vector" as a replacement

### PR DESCRIPTION
**Pull request by immerrr**
This PR enables back raster resizing that was replaced with "vector" one in 97f59adad2642d1e552206a3b0e95d9969b9673d, because of a seemingly unsolvable issue described [here](https://github.com/scrapinghub/splash/issues/169#issuecomment-71326285).

* new `scale_method` parameter added to HTTP & Lua API with values `raster` (default) and `vector`

* PNG rendering has undergone refactoring that among other things paves the path to "clip_rect" functionality.

Benchmark results are available [here](https://gist.github.com/immerrr/8f996b10f7a4d273566e). In short:
- this implementation of raster resize is faster than the pre-32k-fix because of decreased PNG compression level and some cropping of rendering area
- old vector-based resize, as expected, runs circles around raster-based, both in terms of CPU and RSS
but the two implementations are on par in no-resize and resized-image-heavy-site cases

### CODE REVIEW

File: splash/tests/test_render.py [hunk 1]
```diff
@@ -589,6 +586,18 @@ def _do_same_requests(self, url, params, other_format):
         return r1, r2
 
 
+class RenderVectorPngTest(RenderPngTest):
+    def request(self, query, *args, **kwargs):
+        query.setdefault('scale_method', 'vector')
+        return super(RenderVectorPngTest, self).request(query, *args, **kwargs)
+
+
+class RenderVectorJsonTest(RenderJsonTest):
```
Comment (kmike) [inline splash/tests/test_render.py]: Why is it necessary to run render.json tests with scale_method=vector, what do they check which is not checked in render.png tests?

Comment (immerrr) [inline splash/tests/test_render.py]: I was trying to ensure that scale_method argument is actually passed through.

File: splash/qtutils.py [hunk 2]
```diff
@@ -254,171 +256,380 @@ def swap_byte_order_i32(buf):
...
+    logger.log("png render: rendering %s of the web page" % web_rect,
+               min_level=2)
+    logger.log("png render: rendering into %s of the canvas" % render_rect,
+               min_level=2)
+    logger.log("png render: canvas size=%s" % canvas_size, min_level=2)
+
+    to_paint = render_rect.intersected(QRect(QPoint(0, 0), canvas_size))
+    # QPainter has several issues when its rendering area is more than 2**15 in
+    # any dimension.
+    need_tiling = max(to_paint.width(), to_paint.height()) >= (1 << 15)
```
Comment (kmike) [inline splash/qtutils.py]: `2**15` reads better :)

Also, the comment says "more than", while the code uses ">=". I don't care about it though, ">=" is fine to be on a safe side.

File: splash/qtutils.py [hunk 3]
```diff
@@ -254,171 +256,380 @@ def swap_byte_order_i32(buf):
     return arr.tostring()
 
 
+def render_qwebpage(web_page, logger=None, width=None, height=None,
```
Comment (kmike) [inline splash/qtutils.py]: What do you think about moving all render-related functions and classes into a dedicated module? Something like qtutils_render.py.

**Comment by kmike** [comment]
The code looks good, waiting for the benchmark results.

File: splash/qtutils.py [hunk 4]
```diff
@@ -254,171 +256,380 @@ def swap_byte_order_i32(buf):
...
+        logger.log("png render: image is trimmed vertically")
+        hcut = image_size.height()
+        if in_viewport.size() == out_viewport.size():
+            # If no resizing then we need to render exactly the requested
+            # number of pixels vertically.
+            hrender = hcut
+        else:
+            # If there's a resize, there will be interpolation.  We must ensure
+            # that both pixels that contribute to the color of the last row of
+            # the resulting image are rendered.
```
Comment (kmike) [inline splash/qtutils.py]: What if more than 2 pixels contribute to the last row? is the formula above still valid?

Comment (immerrr) [inline splash/qtutils.py]: No more than 2 pixels in any dimension contribute to the value of an output pixel in bilinear interpolation algorithm. Bicubic or Lanczos filters have bigger windows, so we'll need to update the formula accordingly before using one of those.

Comment (immerrr) [inline splash/qtutils.py]: If the upscaled pixel lands on the original one perfectly, of course, there will be no second pixel, but this should be handled by this formula. I'm going to test this, just to be sure.

File: splash/qtutils.py (hunk 2 above)
Comment (immerrr) [inline splash/qtutils.py]: Turns out, 32766 is the last safe value, updated the code.

**Comment by immerrr** [comment]
I've updated the first message with [the results of the benchmark](https://gist.github.com/immerrr/8f996b10f7a4d273566e). In short:
- this implementation of raster resize is faster than the pre-32k-fix because of decreased PNG compression level and some cropping of rendering area
- old vector-based resize, as expected, runs circles around raster-based, both in terms of CPU and RSS
but the two implementations are on par in no-resize and resized-image-heavy-site cases

**Comment by kmike** [comment]
Looks good!

**Pull request merged**